File: packages/ts/cairn-p2p/src/mesh/relay.ts
// Mesh relay forwarding (spec/09 section 9.3)
//
// Handles forwarding opaque encrypted bytes between peers. Relay peers
// cannot read, modify, or forge relayed content.

import { CairnError } from '../errors.js';
import type { MeshConfig } from './index.js';

/** Unique identifier for a relay session. */
export type RelaySessionId = number;

/** A relay session bridging two peers through this node. */
export interface RelaySession {
  /** The unique session identifier. */
  id: RelaySessionId;
  /** The source peer ID (hex, requesting the relay). */
  source: string;
  /** The destination peer ID (hex, being relayed to). */
  destination: string;
}

/**
 * Manages relay sessions for this peer.
 *
 * Enforces relayWilling and relayCapacity from MeshConfig.
 * Each relay session bridges two streams, forwarding opaque bytes.
 */
export class RelayManager {
  private _config: MeshConfig;
  private readonly _sessions = new Map<RelaySessionId, RelaySession>();
  private _nextSessionId = 1;

  constructor(config: MeshConfig) {
    this._config = { ...config };
  }

  /**
   * Request to start a new relay session.
   *
   * Validates that this peer is willing to relay, has capacity,
   * and the destination is not the source.
   */
  requestRelay(source: string, destination: string): RelaySessionId {
    if (!this._config.meshEnabled) {
      throw new CairnError('MESH_DISABLED', 'mesh routing disabled');
    }

    if (!this._config.relayWilling) {
      throw new CairnError('RELAY_NOT_WILLING', 'relay not willing');
    }

    if (this._sessions.size >= this._config.relayCapacity) {
      throw new CairnError(
        'RELAY_CAPACITY_FULL',
        `relay capacity full (${this._sessions.size}/${this._config.relayCapacity})`,
      );
    }

    if (source === destination) {
      throw new CairnError(
        'RELAY_CONNECTION_FAILED',
        'source and destination are the same peer',
      );
    }

    const id = this._nextSessionId++;
    this._sessions.set(id, { id, source, destination });
    return id;
  }

  /** Close a relay session. Returns true if the session existed. */
  closeSession(sessionId: RelaySessionId): boolean {
    return this._sessions.delete(sessionId);
  }

  /** Get the number of active relay sessions. */
  get activeSessionCount(): number {
    return this._sessions.size;
  }

  /** Get a relay session by ID. */
  getSession(sessionId: RelaySessionId): RelaySession | undefined {
    return this._sessions.get(sessionId);
  }

  /** Get the remaining relay capacity. */
  get remainingCapacity(): number {
    return Math.max(0, this._config.relayCapacity - this._sessions.size);
  }

  /** Check whether this peer is willing to relay. */
  get isWilling(): boolean {
    return this._config.relayWilling;
  }

  /** Update the mesh configuration. */
  updateConfig(config: MeshConfig): void {
    this._config = { ...config };
  }
}

File: packages/py/cairn-p2p/src/cairn/mesh/relay.py
"""Relay session management for multi-hop mesh routing."""

from __future__ import annotations

import os
from dataclasses import dataclass

from cairn.mesh.router import (
    MeshConfig,
    RelayCapacityFullError,
    RelayNotWillingError,
)


@dataclass
class RelaySession:
    """An active relay session forwarding data between two peers."""

    session_id: bytes
    source: bytes
    destination: bytes
    bytes_relayed: int = 0

    def record_forwarded(self, nbytes: int) -> None:
        self.bytes_relayed += nbytes


class RelayManager:
    """Manages relay sessions for a mesh-enabled peer."""

    def __init__(self, config: MeshConfig | None = None) -> None:
        self._config = config or MeshConfig()
        self._sessions: dict[bytes, RelaySession] = {}

    @property
    def config(self) -> MeshConfig:
        return self._config

    @property
    def active_sessions(self) -> int:
        return len(self._sessions)

    def create_session(
        self, source: bytes, destination: bytes
    ) -> RelaySession:
        """Create a new relay session.

        Raises RelayNotWillingError or RelayCapacityFullError.
        """
        if not self._config.relay_willing:
            raise RelayNotWillingError()

        if len(self._sessions) >= self._config.relay_capacity:
            raise RelayCapacityFullError(
                len(self._sessions), self._config.relay_capacity
            )

        session_id = os.urandom(16)
        session = RelaySession(
            session_id=session_id,
            source=source,
            destination=destination,
        )
        self._sessions[session_id] = session
        return session

    def get_session(
        self, session_id: bytes
    ) -> RelaySession | None:
        return self._sessions.get(session_id)

    def close_session(self, session_id: bytes) -> None:
        self._sessions.pop(session_id, None)

    def close_all(self) -> None:
        self._sessions.clear()
